### ISSUE TITLE
Auto update hashes for apps that don't post hashes proposal

### ISSUE THREAD

**Issue by nitrocode** [lukesampson/scoop#2942]
What do you think about a github bot that would run on heroku or somewhere, verify the hash of each app, if there is a discrepancy, create a branch, update the hash, commit, push, and submit a pr for review, then continue until all apps are checked?

This would prevent apps from having to be manually updated when they become uninstallable once hash conflicts occur.

Once the bot is trusted enough, it could even auto merge the changes.

**Comment by r15ch13** [comment]
This is pretty much what my update service is doing. Would require `bin/auto-pr.ps1` to run `bin/checker.ps1` with the -force flag.

https://github.com/r15ch13/Excavator

**Comment by nitrocode** [comment]
@r15ch13 wow! Where is this hosted? does it work? how often does it run? are there any stats to see when it found a conflict and when the hash was updated / merged in?

I'm happy to hear that I wasn't the only one with that idea.

**Comment by r15ch13** [comment]
@nitrocode it's running on my personal root server (hosted by hetzner.de) every hour (see [docker-compose.yml](https://github.com/r15ch13/Excavator/blob/master/docker-compose.yml) for all the current settings).
Currently there is no way to see the status, but I could publish the log output of `checkver.ps1`, which is already stored for each bucket.

**Comment by nitrocode** [comment]
Interesting. If this is running every hour, then how come there are many hashes that have been out of date for a little while? For instance the cygwin hashes were out of date for longer than an hour.

Yes, an output log would be very useful, even if it's committed to a separate excavator-log repo. 😄

**Comment by r15ch13** [comment]
That's because `checkver.ps1` is not run with the `-force` flag.
This would cause a big traffic hit for many project sites (and my server) because it downloads all files every time. Not so much a problem for projects that provide checksums. But, from my experience, these can be wrong from time to time too.

**Comment by nitrocode** [comment]
Thank you for sending this! So is this running for informational purposes or eventually will there be some kind of allow list where some jsons can auto update hashes and troublesome ones have to be manually updated?

**Comment by r15ch13** [comment]
@nitrocode These auto-updates are running right now.
Every commit like `app: Update to version *` you see on https://github.com/lukesampson/scoop/commits/master is produced by that service.

A list where someone can add a program to force an update on the next cycle or checking the Github API for new issues opened containing "hash check failed" would be nice, but this has no priority yet.

**Comment by rasa** [comment]
How about instead of downloading the file each time, the script could do a HEAD request, and check the last-modified and etag headers, a la https://github.com/rasa/scoops/blob/master/check_https.py#L265 , and only download if needed.

**Comment by Ash258** [comment]
Already reported hash fails could be automated via [Github Actions](https://github.com/features/actions). It should be rolled from beta in few weeks.

[Example action](https://github.com/jessfraz/branch-cleanup-action)

# TL;DR

It's only simple settings, which would look like this:

```
workflow "Check hash if reported" {
  on = "issue"
  resolves = ["check hashes"]
}

action "check hashes" {
  uses = "some/repository@master"
  secrets = ["GITHUB_TOKEN"]
}
```

And repository is only simple docker repository, with main wrapper Dockerfile, where you can do everything, what you need, in this case, use some pwsh compatible image and create powershell script with these steps:

1. Parse issue title `(?<app>.*)@(?<version>.*) hash check failed`
1. Call check-hashes with given app
1. If there is change
    1. Update
    1. Commit with closing directive
1. If there is no change
    1. Comment on issue, that everything is allright
